### PR TITLE
[LAM-82] API calls to start and stop flink applications

### PR DESCRIPTION
**Pull request by ioantsaf**
API calls to start and stop a deployed flink application on a lambda instance.
Ansible and supervisord are used to start/stop the applications.

### CODE REVIEW

File: ansible/roles/common/files/path.sh [hunk 1]
```diff
@@ -0,0 +1 @@
+export PATH=$PATH:/usr/local/flink/bin
```
Comment (gouzouni625) [inline ansible/roles/common/files/path.sh]: I think it is a bad practice to hard code the installation path (usr/local) of Apache Flink. Have you considered using the environment template file the common role already deployes?

https://github.com/grnet/okeanos-LoD/blob/master/ansible/roles/common/templates/environment.j2

Comment (ioantsaf) [inline ansible/roles/common/files/path.sh]: Yes, the /etc/environment file does not work for setting the $PATH, as it
gets overridden by the $PATH set in /etc/profile. I can change the path.sh
file to be a template, and get the path dynamically.

File: webapp/backend/events.py [hunk 2]
```diff
@@ -142,3 +144,23 @@ def delete_lambda_instance_application_connection(lambda_instance_uuid, applicat
...
                                                     application=application).delete()
+
+
+@shared_task
+def start_stop_application(lambda_instance_uuid, application_uuid, action, app_type):
+    lambda_instance = LambdaInstance.objects.get(uuid=lambda_instance_uuid)
+    application = Application.objects.get(uuid=application_uuid)
+    instanceapplication = LambdaInstanceApplicationConnection.objects.\
+        get(lambda_instance=lambda_instance, application=application)
+    if action == 'start':
```
Comment (gouzouni625) [inline webapp/backend/events.py:162]: When changing an object from the database, do `instanceapplication.save()` after all the changes are done, to save it back to the database.

File: webapp/backend/views.py [hunk 3]
```diff
@@ -408,6 +412,129 @@ def withdraw(self, request, uuid, format=None):
...
+        return Response({
+            "status": {
+                'code': status_code,
+                'short-description': ResponseMessages.short_descriptions['application_stop']
+            }}, status=status_code)
+
+    def get_application_instance_connection(self, request, uuid):
+        """
+        This method is used to withdraw an application from a specified lambda instance. Responds
+        to POST requests on url r'^{prefix}/{lookup}/{methodname}{trailing_slash}$'
```
Comment (gouzouni625) [inline webapp/backend/views.py]: Please refactor this method's documentation.

File: webapp/backend/views.py [hunk 4]
```diff
@@ -408,6 +412,129 @@ def withdraw(self, request, uuid, format=None):
...
+        # Check if the specified lambda instance exists.
+        lambda_instances = LambdaInstance.objects.filter(uuid=lambda_instance_uuid)
+        if not lambda_instances.exists():
+            raise CustomNotFoundError(CustomNotFoundError.messages['lambda_instance_not_found'])
+        lambda_instance = lambda_instances[0]
+
+        # Check if the specified application exists.
+        applications = self.get_queryset().filter(uuid=application_uuid)
+        if not applications.exists():
+            raise CustomNotFoundError(CustomNotFoundError.messages['application_not_found'])
```
Comment (gouzouni625) [inline webapp/backend/views.py:548]: There are 2 types of messages, error messages and response messages. Error messages are returned via custom exceptions. There is a number of custom exception types each with its own messages. When throwing a custom exception with a message, make sure you add that message to the `messages` dictionary of the respective exception. ( https://github.com/grnet/okeanos-LoD/blob/devel/webapp/backend/exceptions.py )

File: webapp/backend/views.py [hunk 5]
```diff
@@ -408,6 +412,129 @@ def withdraw(self, request, uuid, format=None):
...
+        # Create a task to stop the application.
+        tasks.start_stop_application.delay(lambda_instance_uuid=lambda_instance_uuid,
+                                           app_uuid=uuid, app_action=app_action, app_type=app_type)
+
+        # Return an appropriate response.
+        status_code = status.HTTP_202_ACCEPTED
+        return Response({
+            "status": {
+                'code': status_code,
+                'short-description': ResponseMessages.short_descriptions['application_stop']
```
Comment (gouzouni625) [inline webapp/backend/views.py:523]: There are 2 types of messages, error messages and response messages. Response messages are returned when no exceptions have occurred. They are saved under the `ResponseMessages` class and are organized in dictionaries. When returning a response message from this class, make sure you add it on a suitable dictionary.
( https://github.com/grnet/okeanos-LoD/blob/devel/webapp/backend/response_messages.py )

**Comment by ioantsaf** [comment]
please review @gouzouni625 @makistsantekidis

**Comment by gouzouni625** [comment]
+1

**Pull request merged**